ignite/pkg/cosmosclient/rpc.go

```diff
@@ -0,0 +1,234 @@
+package cosmosclient
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/tendermint/tendermint/libs/bytes"
+	"github.com/tendermint/tendermint/rpc/client"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+	"github.com/tendermint/tendermint/types"
+)
+
+// rpcWrapper is a rpclient.Client but with more contextualized errors.
+// Useful because the original implementation may return JSON errors when the
+// requested node is busy, which is confusing for the user. With rpcWrapper,
+// the error is prefixed with 'error while requesting node xxx: JSON error'.
+// TODO(tb): we may remove this wrapper once https://github.com/tendermint/tendermint/issues/9312 is fixed.
+type rpcWrapper struct {
+	rpcclient.Client
+	nodeAddress string
+}
+
+func rpcError(node string, err error) error {
```

Comment:
```suggestion
// res is a generic, sorry I don't remember the syntax.
func rpcError(res any, err error) (res any, err error) {
   if err == nil {
       return
   }
   
   err = fmt.Errorf("error while requesting node '%s': %w", node, err
   
   return
}
```


Why not making this a private method of `rpcWrapper`.  so you don't need to provide `node` each time?

Reply:
This my not work tho, just assumed that Go generics support it without hassle.

Reply:
Good idea but I'm also using for `QueryClient`, see https://github.com/ignite/cli/pull/2813/files#diff-3bd1f8e07c32a86f2d79c7726abfa82e730c9d85045c6bd1d53bc151bffa52d0R23

Reply:
About making the method accepting nil error, maybe we can rely on `github.com/pkg/errors` :

```
func rpcError(node string, err error) error {
	return errors.Wrapf(err, "error while requesting node '%s'", node) // returns nil if err is nil
}
```

Reply:
I tried and it doesn't compile :

```go
func (rpc rpcWrapper) rpcError[T any](res T, node string, err error) (T, error) {
	return res, errors.Wrapf(err, "error while requesting node '%s'", rpc.nodeAddress)
}

func (rpc rpcWrapper) ABCIInfo(ctx context.Context) (*ctypes.ResultABCIInfo, error) {
	return rpc.rpcError(rpc.Client.ABCIInfo(ctx))
}
```

Actually, methods are not permitted to have type params not specified on the type of the receiver. And by definition, we can't define that type in the receiver because it can be any of the returned RPC response types.

And personally, I find that code very hard to read ^^

Reply:
@ilgooz the version with pkg/errors [eaaf008](https://github.com/ignite/cli/pull/2813/commits/eaaf0082bb1717a5fb3dc42d87b6e0c4f3b5d50b)